Firmware/ultralcd.cpp

```diff
@@ -1031,11 +1031,11 @@ void lcd_commands()
 			lcd_setstatuspgm(_i("PID cal. finished"));////MSG_PID_FINISHED c=20
 			setAllTargetHotends(0);  // reset all hotends temperature including the number displayed on the main screen
 			if (_Kp != 0 || _Ki != 0 || _Kd != 0) {
-			strcpy(cmd1, "M301 P");
+			strcpy_P(cmd1, PSTR("M301 P"));
 			strcat(cmd1, ftostr32(_Kp));
-			strcat(cmd1, " I");
+			strcat_P(cmd1, PSTR(" I"));
 			strcat(cmd1, ftostr32(_Ki));
-			strcat(cmd1, " D");
+			strcat_P(cmd1, PSTR(" D"));
```

Comment:
Maybe this can be turned into a single sprintf_P. It's just building a command using the resulting PID values. Same for M303 above.

Reply:
I thought about it but didn't have time to test it. `ftostr32` limits the string width, but `sprintf` wouldn't truncate the length.

Reply:
Another note, slightly related to `ftostr32`. I've found that the code in general in conv2str.cpp can be optimised but 80% of the code is unused so perhaps we can just replace it by being a bit more smart with AVR's stdio library https://www.nongnu.org/avr-libc/user-manual/group__avr__stdio.html